### ISSUE TITLE
F.batch_norm output nan

### ISSUE THREAD

**Issue by pzy2000** [pytorch/pytorch#106605]
### 🐛 Describe the bug

import torch
import numpy as np
import torch.nn.functional as F
torch.manual_seed(0)
x_np = np.load('x_before.npy')
x = torch.tensor(x_np).cuda()
print(np.isnan(x.detach().cpu().numpy()).any())
x = F.batch_norm(x, running_mean=torch.randn(x.shape[1]).float().to(x.device),
               running_var=torch.randn(x.shape[1]).float().to(x.device), training=False)
print(np.isnan(x.detach().cpu().numpy()).any())

x_before.npy is provided below
[x_before.zip](https://github.com/pytorch/pytorch/files/12258168/x_before.zip)

with the given input x_before.npy, F.batch_norm outputs nan

### Versions

PyTorch version: 1.10.1+cu111
Is debug build: False
CUDA used to build PyTorch: 11.1
ROCM used to build PyTorch: N/A

OS: Ubuntu 18.04.6 LTS (x86_64)
GCC version: (Ubuntu 7.5.0-3ubuntu1~18.04) 7.5.0
Clang version: Could not collect
CMake version: Could not collect
Libc version: glibc-2.27

Python version: 3.8.13 (default, Mar 28 2022, 11:38:47)  [GCC 7.5.0] (64-bit runtime)
Python platform: Linux-4.15.0-211-generic-x86_64-with-glibc2.17
Is CUDA available: True
CUDA runtime version: Could not collect
CUDA_MODULE_LOADING set to: 
GPU models and configuration: 
GPU 0: Tesla V100-SXM2-32GB
GPU 1: Tesla V100-SXM2-32GB

Nvidia driver version: 470.182.03
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.8.0.5
/usr/lib/x86_64-linux-gnu/libcudnn_adv_infer.so.8.0.5
/usr/lib/x86_64-linux-gnu/libcudnn_adv_train.so.8.0.5
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_infer.so.8.0.5
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_train.so.8.0.5
/usr/lib/x86_64-linux-gnu/libcudnn_ops_infer.so.8.0.5
/usr/lib/x86_64-linux-gnu/libcudnn_ops_train.so.8.0.5
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:        x86_64
CPU op-mode(s):      32-bit, 64-bit
Byte Order:          Little Endian
CPU(s):              10
On-line CPU(s) list: 0-9
Thread(s) per core:  1
Core(s) per socket:  10
Socket(s):           1
NUMA node(s):        1
Vendor ID:           GenuineIntel
CPU family:          6
Model:               85
Model name:          Intel(R) Xeon(R) Gold 6248 CPU @ 2.50GHz
Stepping:            7
CPU MHz:             2499.998
BogoMIPS:            4999.99
Hypervisor vendor:   KVM
Virtualization type: full
L1d cache:           32K
L1i cache:           32K
L2 cache:            4096K
L3 cache:            16384K
NUMA node0 CPU(s):   0-9
Flags:               fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ss ht syscall nx pdpe1gb rdtscp lm constant_tsc arch_perfmon rep_good nopl xtopology cpuid tsc_known_freq pni pclmulqdq ssse3 fma cx16 pcid sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand hypervisor lahf_lm abm 3dnowprefetch topoext cpuid_fault invpcid_single pti ssbd ibrs ibpb fsgsbase tsc_adjust bmi1 hle avx2 smep bmi2 erms invpcid rtm mpx avx512f avx512dq rdseed adx smap clflushopt clwb avx512cd avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves arat umip pku ospke avx512_vnni

Versions of relevant libraries:
[pip3] numpy==1.24.2
[pip3] torch==1.10.1+cu111
[pip3] torchaudio==0.10.1+cu111
[pip3] torchsummary==1.5.1
[pip3] torchvision==0.11.2+cu111
[conda] cudatoolkit               11.1.1               h6406543_8    conda-forge
[conda] numpy                     1.24.2                   pypi_0    pypi
[conda] torch                     1.10.1+cu111             pypi_0    pypi
[conda] torchaudio                0.10.1+cu111             pypi_0    pypi
[conda] torchsummary              1.5.1                    pypi_0    pypi
[conda] torchvision               0.11.2+cu111             pypi_0    pypi

cc @albanD @mruberry @jbschlosser @walterddr @mikaylagawarecki

**Comment by mikaylagawarecki** [comment]
Hey, I see you are using 1.10.1, could you check whether this reproduces on 2.0.1/ the latest main please!

**Comment by pzy2000** [comment]
> Hey, I see you are using 1.10.1, could you check whether this reproduces on 2.0.1/ the latest main please!

thanks for reminding!I just checked it on 2.0.1 stable, it is the same case.

**Comment by tringwald** [comment]
I haven't checked the provided ZIP file, but from the code below I can see that you are passing the output of `torch.randn` to `running_var`. This is problematic as `torch.randn` yields normal distributed values, including negative ones. During the calculation of the batch normalization, these values are passed to `std::sqrt`, which of course does not handle complex number cases for negative inputs. 
Please note, that a negative variance is not a mathematically valid (or reasonable) input.
```
x = F.batch_norm(
  x, 
  running_mean=torch.randn(x.shape[1]).float().to(x.device), 
  running_var=torch.randn(x.shape[1]).float().to(x.device), 
  training=False
)
```

**Comment by pzy2000** [comment]
thanks for your explanation!
